os/fs/smartfs/smartfs_utils.c

```diff
@@ -1065,209 +1163,276 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
-		 * room for the new entry.
-		 */
-
-		if (nextsector == SMARTFS_ERASEDSTATE_16BIT) {
-			/* Allocate a new sector and chain it to the last one */
-
-			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-			if (ret < 0) {
-				goto errout;
-			}
+errout:
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	if (ret != OK) {
+		ret = -ENOENT;
+	}
 
-			nextsector = (uint16_t)ret;
+	if (ret == OK) {
+		RESET_ENTRY_FLAGS(ret_entry);
+	}
 
-			/* Chain the next sector into this sector sector */
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	fdbg("\nfIND_FIRSTINVALIDENTRY returns invalidentry->flags = %x\n", ret_entry->flags);
+	return ret;
+}
 
-			chainheader->nextsector[0] = (uint8_t)(nextsector & 0x00FF);
-			chainheader->nextsector[1] = (uint8_t)(nextsector >> 8);
-			//*((uint16_t *)chainheader->nextsector) = nextsector;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, nextsector);
-			readwrite.count = sizeof(uint16_t);
-			readwrite.buffer = chainheader->nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error chaining sector %d\n", nextsector);
-				goto errout;
-			}
+/****************************************************************************
+ * Name: smartfs_writeentry
+ *
+ * Description: Takes care of the operation of renaming an existing file
+ *              in a particular directory to a new name in the same dircetory
+ *              while minimizing the need to allocate new sectors for
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
 
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-			 /* TODO We will modify whole of create entry to integrity MTD operation soon.
-			  * We will always create sector first, and then update inside of sector. 
-			  * This will help to clear by fs_recover because all sector will become to isolated sector...
-			  * It is temp FIX, If we use sector buffer to support CRC, currently created new parent sector which is chained 
-			  * to previous parent sector is not committed yet. So Now we break loop to prevent read error 
-			  */
-			offset = sizeof(struct smartfs_chain_header_s);
-			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
-			psector = nextsector;
-			break;
-#endif
-		}
+int smartfs_writeentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s new_entry, const char *filename, uint16_t type, mode_t mode, struct smartfs_entry_s *ret_entry)
+{
+	fdbg("\n\n");
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint16_t psector;
+	uint16_t offset;
+	uint16_t entrysize;
+	struct smartfs_entry_header_s *entry;
 
-		/* Now update to the next sector */
+	/* Validate the name isn't too long */
 
-		psector = nextsector;
+	if (strlen(filename) > fs->fs_llformat.namesize) {
```

Comment:
Flow is create or find entry and then write entry right?
Then this should be checked very first time. 
So, start of create or find entry is correct position.
Then we don't have to waste many of procedure until it reaches here with long file name.
If we consider expandability, modulization is most important.

Reply:
Yes, agreed. I will move this check to smartfs_smart.c.
Every function hat involves creation of an entry with a new name will go thorugh this check first.